### PR TITLE
Modify the `PlatformEngines.get_server_dir` function to convert `domain:port` to `domain_port`

### LINKED ISSUE #3130: `PlatformEngines.get_server_dir` does not handle `host:port` correctly on Windows
On Linux ([example log](https://github.com/JuliaLang/Pkg.jl/runs/7241867795?check_suite_focus=true)):

```
┌ Info: 
│   ENV["JULIA_PKG_SERVER"] = "http://localhost:8888"
└   Pkg.PlatformEngines.get_server_dir(ENV["JULIA_PKG_SERVER"]) = "/home/runner/.julia/servers/localhost:8888"
┌ Info: 
│   ENV["JULIA_PKG_SERVER"] = "http://localhost:8888/"
└   Pkg.PlatformEngines.get_server_dir(ENV["JULIA_PKG_SERVER"]) = "/home/runner/.julia/servers/localhost:8888"
```

On macOS ([example log](https://github.com/JuliaLang/Pkg.jl/runs/7241867856?check_suite_focus=true)):

```
┌ Info: 
│   ENV["JULIA_PKG_SERVER"] = "http://localhost:8888"
└   Pkg.PlatformEngines.get_server_dir(ENV["JULIA_PKG_SERVER"]) = "/Users/runner/.julia/servers/localhost:8888"
┌ Info: 
│   ENV["JULIA_PKG_SERVER"] = "http://localhost:8888/"
└   Pkg.PlatformEngines.get_server_dir(ENV["JULIA_PKG_SERVER"]) = "/Users/runner/.julia/servers/localhost:8888"
```

On Windows ([example log](https://github.com/JuliaLang/Pkg.jl/runs/7241867940?check_suite_focus=true)):

```
┌ Info: 
│   ENV["JULIA_PKG_SERVER"] = "http://localhost:8888"
└   Pkg.PlatformEngines.get_server_dir(ENV["JULIA_PKG_SERVER"]) = "localhost:8888"
┌ Info: 
│   ENV["JULIA_PKG_SERVER"] = "http://localhost:8888/"
└   Pkg.PlatformEngines.get_server_dir(ENV["JULIA_PKG_SERVER"]) = "localhost:8888"
```

The master branch of [PkgAuthentication.jl](https://github.com/JuliaComputing/PkgAuthentication.jl) seems to have a version of `get_server_dir` that fixes this bug. I'll investigate further, and I'll try to make a PR.

### PR DESCRIPTION
**Pull request by DilumAluthge**
Fixes #3130

### CODE REVIEW

File: src/PlatformEngines.jl [hunk 1]
```diff
@@ -49,13 +49,19 @@ function get_server_dir(
...
     end
     isempty(Base.DEPOT_PATH) && return
-    joinpath(depots1(), "servers", String(m.captures[1]))
+    m_1 = m[1]::AbstractString
+    if m[2] === nothing
+        return joinpath(Pkg.depots1(), "servers", m_1)
+    else
+        m_3 = m[3]::AbstractString
+        return joinpath(Pkg.depots1(), "servers", m_1 * "_" * m_3)
+    end
```
Comment (pfitzseb) [inline src/PlatformEngines.jl:64]: Thoughts on unconditionally replacing all ["invalid" chars](https://stackoverflow.com/a/31976060/12113178) instead? Would probably end up looking a bit cleaner.

Comment (DilumAluthge) [reply src/PlatformEngines.jl]: Done!

File: src/PlatformEngines.jl [hunk 2]
```diff
@@ -55,7 +55,10 @@ function get_server_dir(
         return
     end
     isempty(Base.DEPOT_PATH) && return
-    joinpath(depots1(), "servers", String(m.captures[1]))
+    m_1 = only(m)::AbstractString
```
Comment (fredrikekre) [inline src/PlatformEngines.jl:58]:
```suggestion
```

File: src/PlatformEngines.jl [hunk 3]
```diff
@@ -55,7 +55,10 @@ function get_server_dir(
...
+    invalid_filename_chars = [':', '/', '<', '>', '"', '/', '\\', '|', '?', '*']
+    dir = join(replace(c -> c in invalid_filename_chars ? '_' : c, collect(m_1)))
```
Comment (fredrikekre) [inline src/PlatformEngines.jl:60]:
```suggestion
    dir = replace(String(only(m)), (c => '_' for c in invalid_filename_chars)...);
```

File: src/PlatformEngines.jl [hunk 4]
```diff
@@ -55,7 +55,9 @@ function get_server_dir(
         return
     end
     isempty(Base.DEPOT_PATH) && return
-    joinpath(depots1(), "servers", String(m.captures[1]))
+    invalid_filename_chars = [':', '/', '<', '>', '"', '/', '\\', '|', '?', '*']
+    dir = replace(String(only(m)), (c => '_' for c in invalid_filename_chars)...);
```
Comment (pfitzseb) [inline src/PlatformEngines.jl:59]: This doesn't work on 1.6 (just saw the backport-1.6 label).

```suggestion
    dir = join(replace(c -> c in invalid_filename_chars ? '_' : c, collect(String(only(m)))))
```
should work, but is also pretty ugly.

**Review by pfitzseb — APPROVED** [review]
LGTM.

**Pull request merged**